resources/templates/md/posts/2016-05-02-multimethods-protocols-explained.md

```diff
@@ -0,0 +1,486 @@
+given musician's name or surname begins with the given letter. Our
+main "algorithm" is in the `pay-tribute` function, which simply
+filters the musicians vector and use `tribute` to polymorphically pay
+tribute to them. It's a very simple function but pretend for a moment
+that the algorithm is actually more complicated. We test the function
+against the `\K` letter and we see the correct result.
+
+But what if we want to add another musician whose role is neither
+`:vocals` nor `:drums`?
+
+```clojure
+(pay-tribute (conj musicians {:name "Jaco Pastorius" :role :bass}) \J)
+;; [...]  No method in multimethod 'tribute' for dispatch value: :bass
+```
+
+And here's the main point of multimethods (and abstractions in
+general): we just need to create another implementation,
+**without modifying** our main algorithm, and everything will work as
+expected.
+
+```clojure
+(defmethod tribute :bass
+  [musician]
+  (str (:name musician) " played the best bass lines ever!"))
+
+(pay-tribute (conj musicians {:name "Jaco Pastorius" :role :bass}) \K)
+;; Jim Morrison had such a perfect voice!
+;; John Bonham was really great at drums! Just like a metronome!
+;; Jaco Pastorius played the best bass lines ever!
+```
+
+We can also create a default implementation, for example we might
+still need a bit of time to come up with a good tribute quote for 
+guitar players.
+
+```clojure
+(defmethod tribute :default
+  [musician]
+  (str (:name musician) " was very talented."))
+
+(pay-tribute (conj musicians {:name "Jimi Hendrix" :role :guitar}) \J)
+;; Jim Morrison had such a perfect voice!
+;; John Bonham was really great at drums! Just a metronome!
+;; Jimi Hendrix was very talented.
+```
+
+You can see how easy is to provide new implementations for different 
+roles, or in other words to extend the algorithm with new
+behaviors -- again: without having to modify any existing code. Before
+we cover protocols few notes:
+
+* `(defmulti tribute (fn [musician] (:role musician)))` could be
+  written as `(defmulti tribute :role)` since keywords are functions.
+* This is a simple dispatch function but could be as complicated as we
+  want, for example we might dispatch not only by role but also by
+  genre: `(fn [musician] [(:role musician) (:genre musician)])`. 
+  Actually, the dispatching function can return any value using all
+  or some of its arguments.
+* In case `:default` means something in your domain logic you can choose
+  another default key in the `defmulti` definition: 
+  `(defmulti name dispatch-fn :default default-value) ` 
+* We can dispatch by the type of the argument (`class`) similarly to
+  what we do in OOP (aka _polymorphism_).
+* For more advanced topics (`isa?`, `derive`, `prefer-method`) see
+  [the official docs](http://clojure.org/reference/multimethods) and
+  relative chapter in
+  [Programming Clojure](https://pragprog.com/book/shcloj2/programming-clojure).
+* Multimethods are rarely used but you can see an interesting use-case in
```

Comment:
Maybe we need to say that they are very slow as they are fully dynamic (the function will always be evaluated and never aot-ed)